### ISSUE TITLE
Viewer: change camera to get other interaction modes (e.g. flight mode)

### ISSUE THREAD

**Issue by nmellado** [STORM-IRIT/Radium-Engine#459]
For some application, it would be interesting to switch from the trackball interaction mode to a flight mode, where the camera is moved in the scene using up/down/left/right keys and the mouse. 

From what I see, this would require to write a new class 
```c++ 
class FirstPersonCamera : public CameraInterface { /* ... */};
```
Then, the problem comes from `Viewer`, which correctly holds a `unique_ptr<CameraInterface>`, but in the implementation i also explicitly calls methods of `TrackballCamera`:
https://github.com/STORM-IRIT/Radium-Engine/blob/e39e9a7440ae1f7bebcb8adea9b55e7bc23fa19d/src/GuiBase/Viewer/Viewer.cpp#L66
https://github.com/STORM-IRIT/Radium-Engine/blob/e39e9a7440ae1f7bebcb8adea9b55e7bc23fa19d/src/GuiBase/Viewer/Viewer.cpp#L367
(this call to `TrackballCamera::getContext()` is repeated multiple times).

-------------------------
So my question is: how do we modify `Viewer` so the type of camera can be easily changed ?
Do you see any other implication to this proposal ?

**Comment by dlyr** [comment]
We can easily use an active handler instead of active context (as a function ptr to the active handleMouse*Event)
The question is also does the trackball camera need to be different from the FirstPersonCamera (or it can be different action on the same camera ... just change the name)

**Comment by nmellado** [comment]
> We can easily use an active handler instead of active context (as a function ptr to the active handleMouse*Event)

ok.

> The question is also does the trackball camera need to be different from the FirstPersonCamera (or it can be different action on the same camera ... just change the name)

As far as I understand, in the current version a `Camera` is more an interaction mode than something else. So it makes sense to create a new class. But it is true that the name is maybe wrong, and that all the code related to the actual camera could be move to a unique `Camera` class, which includes a polymorphic object dealing with the interaction.

**Comment by dlyr** [comment]
Let's say when you have the FirstPerson camera, with interaction, I will check how to properly have both camera interaction/keymaping in viewer.

**Comment by MathiasPaulin** [comment]
I think that a better way to solve this is to allow camera manipulators (trackball, first person, ...) to register their own keymapping. 
For this, instead of calling a static function in the viewer to register keymaping, we can call a virtual method from CameraInterface, overridden in all manipulators, to register the keymapping associated with a manipulator.
This will imply that `Viewer::setupKeyMappingCallbacks`is no more a static method to access the camera manipulator of the viewer. This will also allow applications to manage several independant viewers. As I have a branch with issue #181 fixed (that is somehow related to the camera manipulation), I'l try this and PR as soon as it is functional for comments.

**Comment by MathiasPaulin** [comment]
Solved by #465
